master/internal/api_trials.go

```diff
@@ -42,53 +50,135 @@ func (a *apiServer) TrialLogs(
 	req *apiv1.TrialLogsRequest, resp apiv1.Determined_TrialLogsServer) error {
 	if err := grpc.ValidateRequest(
 		grpc.ValidateLimit(req.Limit),
+		grpc.ValidateFollow(req.Limit, req.Follow),
 	); err != nil {
 		return err
 	}
+
 	_, total, err := trialStatus(a.m.db, req.TrialId)
 	if err != nil {
 		return err
 	}
 
-	offset := api.EffectiveOffset(int(req.Offset), total)
+	filters, err := constructTrialLogsFilters(req)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, fmt.Sprintf("unsupported filter: %s", err))
+	}
 
-	if limit := int32(total - offset); !req.Follow && (limit < req.Limit || req.Limit == 0) {
-		req.Limit = limit
+	logID := int32(-1) // WebUI assumes logs are 0-indexed
```

Comment:
```suggestion
	logID := int32(-1) // WebUI assumes logs are 0-indexed.
```
and on a separate line.